cmd/src/dump.rs

```diff
@@ -0,0 +1,108 @@
+use engine_rocks::{
+    self,
+    raw::{DBOptions, Env},
+    RocksEngine,
+};
+use engine_traits::{Iterable, RaftEngine, RaftLogBatch, CF_DEFAULT};
+use kvproto::raft_serverpb::RaftLocalState;
+use protobuf::Message;
+use raft::eraftpb::Entry;
+use std::fs;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+/// Check the potential original raftdb directory and try to dump data out.
+///
+/// Procedure:
+///     1. Check whether the dump has been completed. If there is a flag file,
+///        delete the original raftdb safely and return.
+///     2. Scan and dump raft data into raft engine.
+///     3. Create a flag file to indicate that dump operation is done.
+///     4. Delete the original raftdb safely.
+pub fn check_and_dump_raft_db<E: RaftEngine>(path: &str, engine: &E, env: Arc<Env>) {
+    if !RocksEngine::exists(path) {
+        let flag_path = generate_flag_path(path);
+        let _ = fs::remove_file(flag_path);
+        return;
+    }
+    let mut opt = DBOptions::default();
+    opt.set_env(env);
+    let db = engine_rocks::raw_util::new_engine_opt(path, opt, vec![])
+        .unwrap_or_else(|s| fatal!("failed to create origin raft engine: {}", s));
+    let origin_engine = RocksEngine::from_db(Arc::new(db));
+    if check_exist_flag_file(path) {
+        delete_safely(path);
+        return;
+    }
+    // Scan raft log from original raftdb and dump into raft engine.
+    let start_key = keys::REGION_RAFT_MIN_KEY;
+    let end_key = keys::REGION_RAFT_MAX_KEY;
+    let mut log_batch = engine.log_batch(0);
+    let mut count_entry = 0;
+    let mut count_region = 0;
+    origin_engine
+        .scan_cf(CF_DEFAULT, start_key, end_key, false, |key, value| {
+            let res = keys::decode_raft_key(key);
+            match res {
+                Err(_) => Ok(true),
+                Ok((region_id, suffix)) => {
+                    match suffix {
```

Comment:
Is the key format the same for raftdb and raft-engine? If it is the same, why bother to parse the keys?

Reply:
No, it's different.